syntax                           = "proto3";
package raftservice;

import "eraftpb.proto";

service RaftService {
  rpc RequestId(IdRequestArgs) returns (IdRequestResponse) {}
  rpc ChangeConfig(eraftpb.ConfChangeV2) returns (ChangeConfigResponse) {}
  rpc SendMessage(eraftpb.Message) returns (RaftMessageResponse) {}
  rpc RerouteMessage(RerouteMessageArgs) returns (RaftMessageResponse) {}
}

message Proposal {
  bytes inner                    = 1;
}

message Entry {
  uint64 key                     = 1; 
  string value                   = 2;
}

// SendMessage

message RaftMessageResponse {
  bytes data                     = 1;
}

// ChangeConfig

enum ChangeConfigResult {
  ChangeConfig_Success                    = 0;
  ChangeConfig_WrongLeader                = 1;
  ChangeConfig_TimeoutError               = 2;
  ChangeConfig_UnknownError               = 3;
}

message ChangeConfigResponse {
  ChangeConfigResult result      = 1;
  bytes data                     = 2;
}

// IdRequest

enum IdRequestResult {
  IdRequest_Success              = 0;
  IdRequest_Error                = 1;
  IdRequest_WrongLeader          = 2;
}

message IdRequestResponse {
  IdRequestResult result         = 1;
  bytes data                     = 2;
}

message IdRequestArgs {
  string addr                    = 1;
}

// Reroute

enum RerouteMsgType {
  ConfChange                     = 0;
  Propose                        = 1;
}

message RerouteMessageArgs {
  bytes proposed_data            = 1;
  eraftpb.ConfChangeV2 conf_change = 2;
  RerouteMsgType type            = 3;
}
